### ISSUE TITLE
Display text earlier 

### ISSUE THREAD

**Issue by rik** [mozilla/addons-frontend#10473]
## Issue
Text can be the LCP for some pages, depending on the content and the resolution. Without access to metrics, I'm not sure how many cases that could represent but might be worth investigating. I think all mobile visits would fall in that category. And that's 46% of the page views in your CrUX report.

Here's a dramatic example: [Best connection ever, from Australia, on a 1024×768 screen, viewing the French version](https://www.webpagetest.org/video/compare.php?tests=210422_AiDcF3_ac8c3e6013f526e3d3cda61900fcdd54-r%3A3-c%3A0&highlightLCP=1&sticky=1&thumbSize=200&ival=100&end=visual)

The images are pushed down because of the resolution but also because the Call To Action button is super wide, making the summary narrow and longer.

Even though the font file finished downloading around 1.5s, the text is not displayed for another 800ms. That's because the main thread is being blocked by JS execution and then some layout.

## Solutions
Font loading strategies all have annoying trade-offs so there's no great solution.

### Preload
By preloading a font, you increase the odds it will be ready by the time FCP happens. But which font should be preloaded: Inter Var or Inter ? I don't think there's a media query that can tell you about variable font support. So some people will download an extra font for nothing. There's also no guarantee that the font will be ready by FCP time. And that also uses more bandwidth which might delay the CSS and the image that could be the LCP, especially because it is a giant font (≥200KB when most preloaded fonts are ≤50KB).

### font-display
By using values `swap`, `optional` or `fallback`, you allow text to be displayed at FCP time which will then also become the LCP time. But the text displayed will be using the fallback font and there's a tradeoff between displaying some useful early and only displaying something that is the final version.

In general, my preference is to use `optional` or `fallback` to give a chance for the font to load early enough but not surprise the user by having the content jump around if the font takes an unusually long time to load. But since you are an SPA, that would mean that it uses the fallback font for the whole session (whereas a regular navigation would get Inter on the second page). So I'd recommend using `swap`.

I have not tested that LCP happens as soon as the text is displayed with the fallback font (I couldn't because of CSP restrictions). But that's my understanding of what [I've read about LCP](https://github.com/WICG/largest-contentful-paint#paint-first-paint).

### Recommandation
Use `font-display: swap` to get the LCP benefit for sure. Preloading the font could damage the LCP score by using too much bandwidth.

**Comment by rik** [comment]
I see there were some recent changes in this area: https://github.com/mozilla/addons-frontend/issues/10222

**Comment by diox** [comment]
Yeah, when we switched from Fira to Inter, we initially had `font-display: swap`, but the flash of content as the font loaded was deemed too brutal to apply for everyone for the whole website.

I don't think we can preload the font, not without completely changing the strategy to make it lighter (https://github.com/mozilla/addons-frontend/issues/10290) and even then, I don't think we want to prioritize the font on every page.

One possible idea is to have 2 different `@font-face`, one with `font-display: swap` and one without, and use the one without on low resolutions. But we haven't looked in the actual feasibility of this yet.

**Comment by diox** [comment]
This is affecting at least:
- Add-on detail page depending on description & window size
- [Add-on reviews page](https://github.com/mozilla/addons-frontend/issues/10445)
- [Add-on versions page](https://github.com/mozilla/addons-frontend/issues/10446)
- [Language Tools page](https://github.com/mozilla/addons-frontend/issues/10444)

https://github.com/mozilla/addons-frontend/pull/10475 would fix it only for "small" screens, at the expense of a flash of content when the font is downloaded. Alternative solutions:
- `font-display: swap` everywhere, not just small screens. 
  - **Pros**: Straightforward, same end result as now once loaded.
  - **Cons**: flash of content on all pages for all screens
- Selective `font-display: swap`. 
  - **Pros**: more limited impact, same end result as now once loaded. 
  - **Cons**: still some flash of content, so doesn't necessarily look better, and more cumbersome for us as we'd have to check every page separately
- Make a small font file with just the essentials for given locales, and preload that on each page. 
  - **Pros**: No flash of content, same end result as now once loaded. 
  - **Cons**: In 99% of cases part of the text that we want to display first to improve our LCP (and user perception in general) is dynamic and doesn't necessarily match the locale of the page, so it's not 100% foolproof while being complex to set up.
- Drop `Inter` completely, maybe add different font for headings. 
  - **Pros**: lightest solution, guaranteed to be faster because no font to load, no flash of content. 
  - **Cons**: looks slightly different (and may differ per OS), need to go over font sizes again

**Comment by rik** [comment]
One thing to note is that a lot of visitors are probably not seeing Inter already (or a mix of Inter and their OS font). That's because [Inter has no glyphs for Arabic or any Asian language](https://rsms.me/inter/glyphs/).

**Comment by diox** [comment]
> Make a small font file with just the essentials for given locales, and preload that on each page. 

We're going to try this first. Investigating subsetting our font for https://github.com/mozilla/addons-frontend/issues/10290 I found out that I can reduce it to 35 KB by only including latin glyphs. As the comment above mentions, for many languages `Inter` was already not used or mixed with an OS font, so it should hopefully be a relatively transparent change.

**Comment by rik** [comment]
Reminder that there is no way to detect variable font support while preloading. But it seems that every browser that supports preloading also supports variable fonts. So it should be alright in practice.

And preloading might hurt the LCP score for pages where an image is the LCP.

**Comment by diox** [comment]
> And preloading might hurt the LCP score for pages where an image is the LCP.

Yes, that's why I want to make sure the font isn't too big.

**Comment by diox** [comment]
My long-term plan for the fonts: in every top-10 locale supported by Inter (\*), generate a minimal subset for that locale that is preloaded when serving requests with that locale in the URL. Those minimal subsets are also declared in the CSS: they have different `unicode-range` set, that can overlap, as long as the smallest subset are defined _last_. The "full" 200 KB (*) font continue to be available and used automatically by the browser if the content somehow doesn't have everything it needs from our minimal subsets.

With that logic, it means that someone browsing /en-US/ content will preload the English minimal subset in any case. If they encounter french content, the French minimal subset will be loaded automatically, if they encounter German content, the German subset will be loaded, and so on. If they encounter content that is not part of a minimal subset they'll automatically download and use the full font. If the first thing they do is load content needing the full font, they will have preloaded a minimal subset for nothing, but that's why it has to be kept small (the exact size depends on the language. The English one should be around 20 KB).

I'm voluntarily ignoring italics for the moment, because there is another issue about it. Italic content will continue to download the separate but full italic font for now.

In the short term, this week I'd like to start with English only minimal subset, but I will follow up with the rest of the subsets the week after.

(\*) I also plan to see what we could strip from the 200 KB file. It contains a lot of glyphs and features we don't necessarily need.

In terms of implementation:
- The subsets have to be generated from the variable font with `pyftsubset Inter-roman.var.woff2 --output-file="Inter-roman-minimal-subset-<lang>.var.woff2 --flavor=woff2 --layout-features="kern" --no-hinting --unicodes="<unicode-range for that lang>"`.
- Unicode ranges will be determined with using https://character-table.netlify.app/ or similar, as well as our own content (`glyphhanger` can be used for this, but there will be manual tweaking involved in each case to ensure the font files makes sense and are kept small while still working well with the preload strategy)
- Fonts will need to be referenced in the CSS file from biggest range to smallest.
- We need to be able to figure out which subset to load from our nodejs code in order to include a preload link. That likely mean there will be some hardcoding somewhere of the font filename.

**Comment by eviljeff** [comment]
re: the subsets for each locale, given there are few character differences between most European languages, will producing a separate file for German for it's extra 4 characters (for example) be worth it? (I'm assuming the file sizes won't be significantly higher, <5k)

**Comment by diox** [comment]
Probably not, that's something I plan to look at as well. I haven't tested them yet, but yeah, I've generated subsets for all major European languages and they are fairly similar, so maybe we'll want to group them. I think English should be kept separate though, as it's the smallest of them all.

**Comment by diox** [comment]
QA: once this lands, in `en-US` page, a `.woff2` file of around 20KB should be preloaded (the network request should happen fairly early, you can check that with developer tools on a cleared cache). On most `en-US` (*) pages, this should be the only font loaded. Once you start going to other locales or visiting a page that happens to use fancy characters or characters from other languages, you should see a second font being loaded (which should be around 200 KB)

Please note the URL of pages that do not behave as expected, and please test on both Chrome and Firefox.

(*) One notable exception is the Dictionaries and Language Packs page, that's covered by https://github.com/mozilla/addons-frontend/issues/10444).

**Comment by diox** [comment]
Already identified one character I thought I had put in the minimal font that is missing: em dash `—` (U+2014). I put the regular hypen (U+2010) and the en dash (U+2013), and I thought I had this one in as well, but I must have missed it. I'll patch this, in the meantime pages using that character will load the second font.

**Comment by diox** [comment]
In addition the preload seems to be missing even though it's working locally and tests pass.

**Comment by AlexandraMoga** [comment]
@diox I'm guessing the testing steps mentioned in https://github.com/mozilla/addons-frontend/issues/10473#issuecomment-827907471 are no longer applicable since we now have the `Inter-roman-subset-en_de_fr_ru_es_pt_pl_it.var.[###].woff2` font preload. 

I believe that the test results outlined in https://github.com/mozilla/addons-frontend/issues/10495#issuecomment-831923779 are also covering this issue. Please correct me if you think otherwise.
